CCE-GH executable

moxcodes opened the pull request:
## Proposed changes

Add the combined GH-CCE executable. This also refactors the GH executable into a 'base' metavariables to avoid code duplication as much as possible.

### Types of changes:

- [ ] Bugfix
- [x] New feature
- [ ] Refactor

### Component:

- [x] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests and `clang-tidy`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### Dependencies

- [x] #2322
- [x] #2319
- [x] #2555
- [x] #3272
- [ ] #3421 
- [ ] #3422

moxcodes committed:
Add combined GH-CCE executable

moxcodes commented:
The CI failure here is most likely due to running out of memory on the github CI virtual machines, so this PR will have to wait until the `DataBox` simplifications are complete so that the build is less expensive.

geoffrey4444 commented:
Would anyone object to squashing this one?

geoffrey4444 commented:
Could you please fix the conflicts?

moxcodes commented:
rebased on develop, fixed all the conflicts.
Cheers!

moxcodes commented:
Rebased on develop and fixed conflicts

moxcodes committed:
Use ThetaVariesFastest in interp. target overload

- previous alterations only fixed one of the function overloads, causing
  inconsistent target points during some executions.

moxcodes commented:
Now dependent on #2555, but the good news is that the optimizations in that PR are enough to get CI to run to completion on this one.

moxcodes commented:
Rebased on develop, no longer dependent.
Cheers!

kidder commented:
@geoffrey4444 @markscheel please look at

kidder commented:
@geoffrey4444 @markscheel please look at

@moxcodes this needs a rebase

moxcodes commented:
rebased
Cheers!

geoffrey4444 commented on src/Evolution/Executables/GeneralizedHarmonic/EvolveGeneralizedHarmonic.hpp:
```diff
@@ -10,11 +10,28 @@
-    : public GeneralizedHarmonicTemplateBase<
-          EvolutionMetavars<InitialData, BoundaryConditions>>,
-      public virtual GeneralizedHarmonicDefaults {
+    : public virtual GeneralizedHarmonicDefaults,
```
Just so that I understand, what does the virtual keyword do here?

moxcodes replied:
Oh, yeah, that's a good question, and maybe deserves a comment to clarify. So, this is a multiple inheritance detail, where a class A inherits both from B and C, which both inherit from D (forming a diamond-shaped graph of the inheritance), and then the 'correct' inheritance path is ambiguous with direct inheritance. The `virtual` keyword flags this problem for the compiler, and it ensures that any derived class inherits from the class denoted with `virtual` only once.

geoffrey4444 commented on src/Evolution/Systems/Cce/InterfaceManagers/GhLocalTimeStepping.hpp:
```diff
@@ -39,6 +39,9 @@ namespace Cce::InterfaceManagers {
  * `GhLocalTimeStepping::retrieve_and_remove_first_ready_gh_data()`.
+ *
+ * \warning To use this interface manager, the CCE system must evolve using a
+ * local time-stepping compatible stepper, such as `AdamsBashforthN`.
```
Is there no good way to enforce this with an ASSERT or something? Just curious.

moxcodes replied:
Pushed a fixup that adds an assert to an option-creatable tag that's sufficiently related that I think it makes sense.

geoffrey4444 commented on src/Evolution/Systems/Cce/OptionTags.hpp:
```diff
@@ -375,8 +375,9 @@ struct SpecifiedStartTime : Tags::StartTime, db::SimpleTag {
   static constexpr bool pass_metavariables = false;
-  static double create_from_options(const double start_time) noexcept {
-    return start_time;
+  static double create_from_options(
```
So I understand, could you please briefly explain what this does? Am I right that it just means users in the input file may omit SpecifiedStartTime? If so, how does the code figure out what start time to use?

moxcodes replied:
This is just a change to fix a problem that I didn't notice from a previous PR because this tag wasn't constructed from options previously -- the `StartTime` options tag returns a `std::optional`, but the contents here mean that it will error if the user specifies `Auto`, because the start time is required to be specified.

geoffrey4444 commented on src/Evolution/Systems/Cce/InterfaceManagers/GhLocalTimeStepping.cpp:
```diff
@@ -93,6 +93,10 @@ void GhLocalTimeStepping::insert_gh_data(
+  if (times_seen_.count(time_id) != 0) {
+    return;
+  }
+  times_seen_.insert(time_id);
```
So I understand, what happened before you made this change?

moxcodes replied:
This was a peculiar bug where a point on the interpolation surface landed exactly on an element boundary, so two elements both claimed to contain that point. So, multiple elements reported to CCE their time steps, which caused the logic in this function to end up incorrect. The code now ensures that time step ids being reported are unique before continuing with the control flow.

geoffrey4444 commented on src/Evolution/Executables/GeneralizedHarmonic/EvolveGeneralizedHarmonic.hpp:
```diff
@@ -3,431 +3,35 @@
 
 #pragma once
 
```
To confirm, is this commit doing nothing other than moving most of the EvolveGeneralizedHarmonic.hpp contents into EvolveGeneralizedHarmonicBase.hpp, so that you can reuse them in EvolveGeneralizedHarmonicWithCce.hpp?

moxcodes replied:
Yeah, my intent is that this commit does nothing but move code so that the 'boring' part of the PR is easy to distinguish from the new code.

geoffrey4444 commented on src/Evolution/Executables/Burgers/EvolveBurgers.hpp:
```diff
@@ -159,7 +159,7 @@ struct EvolutionMetavars {
       dg::Events::Registrars::ObserveFields<1, Tags::Time, observe_fields,
                                             analytic_solution_fields>,
-      Events::Registrars::ObserveTimeStep<EvolutionMetavars>,
+      Events::Registrars::ObserveTimeStep<system>,
```
Looks reasonable, but could you please tell me why you made this change, so I understand?

moxcodes replied:
We've been favoring actions which depend explicitly on the parts of the metavariables that they really need, rather than templating on the whole metavariables -- this was advice I received on a previous PR from @kidder , but my understanding is that like other kinds of code dependencies, it is preferable for the template dependence to be as narrow as it can be so that dependencies are clearer.

geoffrey4444 commented on src/Evolution/Systems/Cce/Components/CharacteristicEvolution.hpp:
```diff
@@ -168,7 +172,8 @@ struct CharacteristicEvolution {
       Actions::ExitIfEndTimeReached,
-      Actions::ReceiveWorldtubeData<Metavariables>>;
+      Actions::ReceiveWorldtubeData<Metavariables>,
+      ::Actions::Goto<CceEvolutionLabelTag>>;
```
I see what you're doing here, and it looks fine, but could you please briefly explain why you made this change, just so I understand?

moxcodes replied:
When CCE operated purely stand-alone, it worked fine to communicate the first time step of data during initialization, but now that it needs to work with the rest of the parallelization, we can't anticipate that anything will get sent during initialization (it really shouldn't send during initialization anyway, our general rule is that no communication should happen during initialization).

geoffrey4444 reviewed (dismissed):
I really don't have any suggestions to make...as far as I can understand what's happening, everything looks reasonable. I do have some questions of the "please help me understand" variety, but these need not hold up this PR's progress.

I tested this on ocean and was able to run a CCE+AH finder evolution of a perturbed black hole, reading from numeric initial data, on 10 nodes (200 processors). Everything looks reasonable to time t=0.4M.

I'm doing a longer run now, and if I notice anything unusual in the output, I'll let @moxcodes know.

wthrowe commented on src/Evolution/Systems/Cce/OptionTags.hpp:
```diff
@@ -464,13 +465,20 @@ struct GhInterfaceManager : db::SimpleTag {
+               dynamic_cast<LtsTimeStepper*>(time_stepper.get()) != nullptr,
+           "The use of full-step interface managers (which perform "
+           "interpolation and extrapolation between the CCE and GH systems) "
+           "require an LTS-compatible time-stepper.");
```
Do you need LTS, or specifically AB?

Also, this looks like a user error, so ASSERT -> ERROR.

moxcodes replied:
I think the constraint I require is that the times of the steps are strictly increasing, and I had the impression that LTS steppers required the same thing -- if those requirements are different, though, I should just require AB specifically.

Agreed, I'll make it `ERROR` instead.

wthrowe replied:
(This is all somewhat academic, since AB is the only implementation of LtsTimeStepper and there are no plans to add any more in the near future.)  Interesting question.  If you don't care about conservation, you can certainly use non-monotonic steps, but I doubt we will ever add a non-conservative stepper.  The only other LTS ideas I have considered are AB-like, and if they are added they should still work for you, so LTS is probably better here.

But maybe what you really want is `number_of_substeps() == 1`.

moxcodes commented:
rebased on develop, and now with @kidder 's very helpful databox improvements, it should actually compile on CI!
There's some fixups for the changes from reviews and rebase.
Cheers!

kidder commented:
@moxcodes this needs a rebase

moxcodes commented:
Rebased on develop. I've squashed the fixups because the diffs would look terrible given all of the refactoring changes.
I've also taken the opportunity to define the GH+CCE executables with the option to have the GH horizon or not, much like the current GH executable.
Cheers!

moxcodes committed:
Terminate CCE when waiting for worldtube data

The `receive_data` calls appropriately restart the algorithm execution.

This change allows phase changes from quiescence detection due to
inactivity (i.e. no more data being supplied by the GH elements); without
terminating at the stage in which the component waits for data, the Algorithm
generates an error if a phase change occurs while the CCE component is inactive.

We want to permit such phase changes, because the CCE component is in a safe
state for balancing or checkpointing when it is inactive and waiting for data.

moxcodes committed:
Add optional template argument for sublists of StepChoosers

moxcodes committed:
Use LTS in CCE regardless of whether it is used for DG

This is already implicitly assumed because we explicitly invoke
`ChangeStepSize` during CCE evolution.

moxcodes commented:
I've updated this now, and added a handful of commits that were necessary to fix changes in the interface logic from time stepping and interpolator changes. The new commits are:
```
Reorder Register phase to before self-start
Add optional template argument for sublists of StepChoosers
Relax constraints on source tags for Horizon
Use self-start specific interface manager during self-start
Use LTS in CCE regardless of whether it is used for DG
Remove duplicates when adding classes to factory creation 
Change CCE slab size after self start
```
I'm happy to split some of these smaller preparatory commits out into a separate PR if that would be easier for the reviewers -- I just thought I'd put them here to begin with, because they are all related to making the CCE system work in tandem with GH.

moxcodes committed:
Remove duplicates when adding classes to factory creation

moxcodes commented:
We're going to try to incorporate changes from #3355 and #3409 so that the CCE side doesn't need to do quite so much to change slab sizes after self-start.

moxcodes committed:
PR #3422

moxcodes committed:
PR #3421

moxcodes commented:
On the plus side, due to the new dense output features, the CCE-GH executable will now support local time-stepping, and independently varies the step size.
However, it also required some significant refactors, so this PR is dependent again.

kidder commented:
@knelli2 can this PR be closed or is there still something we haven't implemented yet?

knelli2 commented:
We have a GH+CCE executable now. We don't have a BBH+CCE executable yet, but it's unclear if we actually want that (except maybe for CCM). And that would be a whole separate PR anyways. So I'll go ahead and close this PR.
